Sources/NIOCore/ByteBuffer-binaryEncodedLengthPrefix.swift

```diff
@@ -0,0 +1,310 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// Describes a way to encode and decode an integer as bytes
+///
+public protocol NIOBinaryIntegerEncodingStrategy {
+    /// Read an integer from a buffer.
+    /// If there are not enough bytes to read an integer of this encoding, return nil, and do not move the reader index.
+    /// If the the full integer can be read, move the reader index to after the integer, and return the integer.
+    /// - Parameters:
+    ///   - as: The type of integer to be read.
+    ///   - buffer: The buffer to read from.
+    /// - Returns: The integer that was read, or nil if it was not possible to read it.
+    func readInteger<IntegerType: FixedWidthInteger>(
+        as: IntegerType.Type,
+        from buffer: inout ByteBuffer
+    ) -> IntegerType?
+
+    /// Write an integer to a buffer. Move the writer index to after the written integer.
+    /// - Parameters:
+    ///    - integer: The type of the integer to write.
+    ///    - buffer: The buffer to write to.
+    /// - Returns: The number of bytes used to write the integer.
+    func writeInteger<IntegerType: FixedWidthInteger>(
+        _ integer: IntegerType,
+        to buffer: inout ByteBuffer
+    ) -> Int
+
+    /// An estimate of the bytes required to write integers using ths strategy.
+    /// Callers may use this to reserve bytes before writing the integer.
+    /// If the actual bytes used by the write function is more or less than this, it may be necessary to shuffle bytes.
+    /// Therefore, an accurate prediction here will improve performance.
+    var requiredBytesHint: Int { get }
+
+    /// Write an integer to a buffer. Move the writer index to after the written integer.
+    /// Call this function if you have already reserved some capacity for an integer to be written.
```

Comment:
The tone of this sentence seems wrong: it's an instruction to the caller, but the caller is us.